### ISSUE TITLE
DEL key visually deletes two characters instead of one

### ISSUE THREAD

**Issue by HennyH** [neovim/neovim#15020]
- `nvim --version`:
```
NVIM v0.5.0
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe /DWIN32 /D_WINDOWS /W3 -DNVIM_TS_HAS_SET_MATCH_LIMIT /MD /Zi /O2 /Ob1 /DNDEBUG /W3 -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -DWIN32 -D_WIN32_WINNT=0x0600 -DINCLUDE_GENERATED_DECLARATIONS -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -ID:/a/neovim/neovim/build/config -ID:/a/neovim/neovim/src -ID:/a/neovim/neovim/nvim-deps/usr/include -ID:/a/neovim/neovim/build/src/nvim/auto -ID:/a/neovim/neovim/build/include
Compiled by runneradmin@fv-az152-786

Features: -acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM\sysinit.vim"
  fall-back for $VIM: "C:/Program Files/nvim/share/nvim"

Run :checkhealth for more info
```

- Operating system/version:  Windows 10 Pro 21H1
- Terminal name/version:  Windows Terminal Preview
- `$TERM`: builtin_vtpcon

### Steps to reproduce using `nvim -u NORC`

Create a HTML file with the following contents:

```html
<p style="margin: 2em; text-align: center;">
    <strong style="color: var(--primary-header-color);">
            ⚠️  Due to Coronavirus restrictions Church capacity has been limited to 150 persons in masks per Mass.
            Four meter social distancing is in place.
    </strong>
</p>
```

Open the file with neovim, move to line 3 and enter INSERT mode. Navigate to the beginning of the "150" on the third line. Press DEL.

I've also asked a question on stackexchange: https://vi.stackexchange.com/questions/31773/delete-key-appears-to-remove-two-characters

### Actual behaviour

Visually both the 1 and the 5 from the 150 are removed *visually*. However if I save the file and view the contents, or ESC into NORMAL mode and then press CTRL+SHIFT+L I can see that only the 1 was deleted.

### Expected behaviour

Visually only the 1 is deleted.

**Comment by rktjmp** [comment]
Can't reproduce this in Linux x Alacritty, xterm or wezterm, probably fair to tag this as a window and/or windows term preview issue?

I assume by the description that you only see the errant behaviour on the line with an emoji on it?

Do you see the same or similar effects on the following example when performing the same steps?

```html
<p style="margin: 2em; text-align: center;">
    <strong style="color: var(--primary-header-color);">
            How does it behave on this line 150 with a tail ⚠️
    </strong>
</p>
```

**Comment by HennyH** [comment]
I paste in the example text:
![image](https://user-images.githubusercontent.com/3156577/125193036-0ce42600-e27d-11eb-8527-11c536941ff0.png)
I then press the DEL key
![image](https://user-images.githubusercontent.com/3156577/125193042-140b3400-e27d-11eb-9abd-384aa27ec2be.png)

So it correctly only removes one character however my cursor ends up in the wrong position (one extra to the right). If I attempt to delete text within the "with" the cursor also moves weirdly in the same manner. 

Interestingly it seems if I instead use Backspace to delete a character I have no issues. `:verbose imap <del>` reports "No mapping found"

**Comment by justinmk** [comment]
what does `:checkhealth nvim` report? specifically the `terminal` section of that report.

**Comment by HennyH** [comment]
When I run `:checkhealth nvim` I only see the following:

![image](https://user-images.githubusercontent.com/3156577/125246582-8c790000-e324-11eb-8b0c-9bfc06ac1cbc.png)

Is there a specific option to get a report relating to the terminal?

**Comment by rktjmp** [comment]
What happens if you paste the line (`⚠️  Due to Coronavirus restrictions Church capacity has been limited to 150 persons in masks per Mass.`) onto the command prompt (outside of vim) and try to delete the 1? There's more going on in nvim, but just to check it's not a basic issue with windows terminal mishandling emoji width.

Or even test it in another editor like nano too, to narrow down the focus.

**Comment by HennyH** [comment]
1. Pasting into commandline prompt and deleting the 1 works as expected (Powershell, Fish under WSL).
2. Pasting the text into nano deletes the 1 as expected.
3. If I open `vi` (`/usr/bin/vi` under WSL) and paste the text in, and then move to before the 1 and press delete it works as expected.

All of the above was done in windows terminal as before.

**Comment by djdv** [comment]
Adding to `vi` in the post above, this also seems to work as expected inside of the [win32 distributions](https://github.com/vim/vim-win32-installer) of `vim` ([v8.2.3161](https://github.com/vim/vim-win32-installer/releases/tag/v8.2.3161)).
`nvim`(v0.6.0-dev+42-geece0735f) seems to corrects itself if I move off of the line after the edit.
You can see characters being drawn into the line number column as well in `nvim`. Probably related to https://github.com/neovim/neovim/issues/14650

https://user-images.githubusercontent.com/13862850/125968652-23353c09-a2f9-46e1-9f24-a2cde7d4ef1b.mp4

[Fullsize video](https://youtu.be/WB0E4EINAyY)

(There seems to be 2 other emoji related issues visible here too, but they're likely unrelated. Probably tied to glyph font fallback. The glyph for `⚠️` is different in `vim` and `nvim`, and a separate glyph `📝` isn't appearing properly in my status bar within `nvim`.)

**Comment by 3N4N** [comment]
Seems like justinmk's intuition was correct. The terminal entry for neovim in Windows and Linux are different.

- Windows:
    ```
    ## terminal
    - INFO: key_backspace (kbs) terminfo entry: ? (not found)
    - INFO: key_dc (kdch1) terminfo entry: ? (not found)
    ```
- Linux:
    ```
    ## terminal
    - INFO: key_backspace (kbs) terminfo entry: key_backspace=\177
    - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
    ```

**Comment by justinmk** [comment]
Assuming this is fixed in recent Windows Terminal, if not leave a comment

**Comment by 3N4N** [comment]
@justinmk Not fixed in Windows Terminal. And behaves the same in the old conhost.exe terminal, too. But this is a conhost/WT bug, not nvim specific. WT+conhost behaves weird with that yellow triangle emoji *outside* nvim, too. So maybe this issue should be on the [microsoft side](https://github.com/microsoft/terminal/issues/4345).

**Comment by zeertzjq** [comment]
Fixed by #19686

**Comment by fatalitirar1** [comment]
i found a solution for a time, may be it can help to find a problem
(it works for me) 
1) if you change keyboard mode to CAPS, backspace start deleting only one character
2) i use kitty terminal, and after i hardcode bakcspace button it starts to work well, code => 

map backspace send_text all \x7f
